server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java

```
@@ -0,0 +1,190 @@
+        // origin: number for long fields; number or string for date fields; string, array, object for geo fields
+        PARSER.declareField(constructorArg(), DistanceFeatureQueryBuilder::originFromXContent,
+            ORIGIN_FIELD, ObjectParser.ValueType.OBJECT_ARRAY_STRING_OR_NUMBER);
+        //pivot: number for long fields; string for dates or geo fields
+        PARSER.declareField(constructorArg(), DistanceFeatureQueryBuilder::pivotFromXContent,
+            PIVOT_FIELD, ObjectParser.ValueType.LONG);
+        declareStandardFields(PARSER);
+    }
+
+    public DistanceFeatureQueryBuilder(String field, Object origin, Object pivot) {
+        this.field = field;
+        this.origin = origin;
+        this.pivot = pivot;
+    }
+
+    public static DistanceFeatureQueryBuilder fromXContent(XContentParser parser) {
+        try {
+            DistanceFeatureQueryBuilder builder = PARSER.apply(parser, null);
+            return builder;
+        } catch (IllegalArgumentException e) {
+            throw new ParsingException(parser.getTokenLocation(), e.getMessage(), e);
+        }
+    }
+
+    public static Object originFromXContent(XContentParser parser) throws IOException {
+        if (parser.currentToken() == XContentParser.Token.VALUE_NUMBER) {
+            return parser.longValue();
+        } else if(parser.currentToken() == XContentParser.Token.VALUE_STRING) {
+            return parser.text();
+        } else if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            return GeoUtils.parseGeoPoint(parser);
+        } else { // if currentToken equals to START_ARRAY)
+            return GeoUtils.parseGeoPoint(parser);
+        }
+    }
+
+    public static Object pivotFromXContent(XContentParser parser) throws IOException {
+        if (parser.currentToken() == XContentParser.Token.VALUE_NUMBER) {
+            return parser.longValue();
+        } else { // if currentToken equals to VALUE_STRING
```

same as above, please check the expected token type and throw error if anything unexpected shows up in the token stream.